layer2-state-provider/history_commitment_provider.go

```diff
@@ -0,0 +1,321 @@
+	idx := 1
+	for _, height := range heights {
+		total := new(big.Int).SetUint64(1)
+		for i := idx; i < len(leafHeights); i++ {
+			total = new(big.Int).Mul(total, new(big.Int).SetUint64(uint64(leafHeights[i])))
+		}
+		increase := new(big.Int).Mul(total, new(big.Int).SetUint64(uint64(height)))
+		opcodeIndex = new(big.Int).Add(opcodeIndex, increase)
+		idx += 1
+	}
+	if !opcodeIndex.IsUint64() {
+		return 0, fmt.Errorf("computed machine start index overflows uint64: %s", opcodeIndex.String())
+	}
+	return OpcodeIndex(opcodeIndex.Uint64()), nil
+}
+
+// Computes the the number of individual opcodes we need to step through a machine at a time.
+// Each challenge level has a different amount of ranges of opcodes, so the overall step size can be computed
+// as a multiplication of all the next challenge levels needed.
+//
+// As an example, this function helps answer questions such as: "How many individual opcodes are there in a single step of a
+// Megastep challenge?"
+func (p *HistoryCommitmentProvider) computeStepSize(challengeLevel uint64) (StepSize, error) {
+	// The last challenge level is over individual opcodes, so the step size is always 1 opcode at a time.
+	if challengeLevel+1 == p.numberOfChallengeLevels() {
+		return 1, nil
+	}
+	// Otherwise, it is the multiplication of all the challenge leaf heights at the next
+	// challenge levels.
+	levels := p.challengeLeafHeights[challengeLevel+1:]
+	total := uint64(1)
+	for _, h := range levels {
+		total *= uint64(h)
+	}
+	return StepSize(total), nil
+}
+
+// Validates a user input must be non-empty and have a size less than the number of challenge levels.
+func (p *HistoryCommitmentProvider) validateStartHeights(
+	startHeights []Height,
+) (validatedStartHeights, error) {
+	if len(startHeights) == 0 {
+		return nil, errors.New("must provide start heights to compute number of hashes")
+	}
+	// A caller specifies a request for history commitments at the first N challenge levels.
+	// This N cannot be greater than the total number of challenge levels in the protocol.
+	deepestRequestedChallengeLevel := len(startHeights) - 1
+	if deepestRequestedChallengeLevel >= len(p.challengeLeafHeights) {
+		return nil, fmt.Errorf(
+			"challenge level %d is out of range for challenge leaf heights %v",
+			deepestRequestedChallengeLevel,
+			p.challengeLeafHeights,
+		)
+	}
+	return validatedStartHeights(startHeights), nil
+}
+
+// A caller specifies a request for a history commitment at challenge level N. It specifies a list of
+// heights at which to compute the history commitment at each challenge level on the way to level N
+// as a list of heights, where each position represents a challenge level.
+// The length of this list cannot be greater than the total number of challenge levels in the protocol.
+// Takes in an input type that has already been validated for correctness.
+func deepestRequestedChallengeLevel(requestedHeights validatedStartHeights) uint64 {
+	return uint64(len(requestedHeights) - 1)
```

Comment:
`requestedHeights` length is 0 then this will underflow

Reply:
It has already been validated for correctness, so we don't need to duplicate the work here. The input type is `validatedStartHeights`, which is known to be nonzero and nonempty